### DISCUSSION TITLE
PID

### DISCUSSION THREAD

**Discussion by jordanpenard** [Zer0-bit/gaggiuino#343] [Ideas]
Hi,
As far as I could see, the control of the heating element is not done via a PID algorithm, it's just a thermostatic control right ?
Is there a technical reason for this or it's just down the list of future improvement ?
I didn't look too much into the code yet, but I'm wondering how much resource handling the pulling of a temperature value and updating the display takes, and if there's enough cpu cycles left to do a correct PID compensation.
Any thought on this ?

**Comment by paintflinger** [comment]
https://discord.com/channels/890339612441063494/922092497847582721/923532894536859669

**Reply by Zer0-bit** [reply]
Not using PID was a deliberate choice that's been made and not related to cpu cycles.

**Reply by jordanpenard** [reply]
Ok understood. How tight do you regulate the temperature at the moment and how much does the temperature ends up fluctuation by ? My only concern at the moment with my stock machine (an old 1st gen GC) is consistency because of the wide variations in temperatures, so I need to always run through a purge routine to catch it on a heating phase to get repeatable pull otherwise it can easily brew way too hot or way too cold. But also it makes it hard to quickly pull back to back shots with good temperature control. In my situation, a PID would help in resolving this issue, and displaying the output power would allow me to know when the boiler has reached enough thermal mass to handle a new pull. 

How do you handle additional features in the code right now ? Would there be a way for me to add such features as an optional feature, and would you be happy to merge such change ? Just to know if it's worth me working on a merge request for this project or if instead I should start fresh somewhere else if it doesn't align with the intent of this project. 

Let me know @Zer0-bit 
Thanks

**Reply by Zer0-bit** [reply]
There' 2 temperatures.
1. Idle - Is controlled to +/- 1C
2. Brew time - Is controlled via applying a delta to compensate for the temp loss during brew (fresh water incoming in the small boiler)

Here's a brew time graph.
![image](https://user-images.githubusercontent.com/42692077/204556450-153c90fd-0c9f-4518-84c3-8d96f802d613.png)

**Reply by Zer0-bit** [reply]
Not interested in PID controls whatsoever so if that's what you're after this ain't for you.

**Reply by jordanpenard** [reply]
Cool, this looks good mate, I'll give it a go, if I can achieve stable temp like this then I might just stick with your control logic, seems like you put some thought into it and it works great.
What about steam temp regulation, I think you just have a single thermocouple at the base of the boiler and just bypassed the steam thermostat, how do you handle steam temp regulation ? 
Thanks a lot for sharing anyway.

**Reply by Sral931** [reply]
Sry for the revival of old stuff, but for other people's reference:

steam temperature is also handled by gaggiuino using the same temperature sensor. The temperature target is simply raised and I think the heater behaves a bit different to give more consistent steam power.

**Comment by bruxy70** [comment]
Still, I wonder, there actually is a Gaggia PID Kit that they sell, and people buy and install.
So is it that this kit is just and useless scam people buy and waste their money with no effect?
Or is the gaggiuino achieving the same result as the PID kit by other means, and so adding PID would have no effect?

I think that for the stick machine the temperature is an issue, and there are workarounds for example for the steaming wand not having enough pressure, by starting the steaming before it turns the boiler off. And I think this one is addressed here...

**Comment by Zer0-bit** [comment]
Everything is handled, temp, steam, pressure, flow, scales...

I won't be speaking about the PID kits as i'm not of great opinion about them also have not tried any so up to the people who tried them to compare.

**Comment by EngMarc** [comment]
Sorry to be dense but what is a PID?
My background is so diverse, I have about 4-5 different meaning for this acronym:   photoionisation detector? Piezo Ionization Sensor? proportional–integral–derivative controller?

No clue in reference to arduino

**Reply by torwag** [reply]
Your last suggestion is it ;)
